src/D2L.Bmx/WriteHandler.cs

```diff
@@ -64,39 +69,50 @@ bool useCredentialProcess
 			using( File.Create( output ) ) { };
 		}
 
-		var data = parser.ReadFile( output );
+		string awsConfigFilePath = useCredentialProcess ? output : SharedCredentialsFile.DefaultConfigFilePath;
+		var awsConfig = GetParsedIniFileOrNull( awsConfigFilePath );
+
+		string awsCredentialsFilePath = useCredentialProcess ? SharedCredentialsFile.DefaultFilePath : output;
+		var awsCredentials = GetParsedIniFileOrNull( awsCredentialsFilePath );
+
 		if( useCredentialProcess ) {
-			string sectionName = $"profile {profile}";
-			if( !data.Sections.ContainsSection( sectionName ) ) {
-				data.Sections.AddSection( sectionName );
+			Debug.Assert( awsConfig is not null );
```

Comment:
What is the thought process behind this being an assertion only in debug builds?

Reply:
The main goal here is to make the compiler's nullable analysis happy in a clear and intentional way.

`awsConfig` should not be null here when `useCredentialProcess` is true (because we'd create the AWS config file above). The compiler cannot infer this, so by default it'd mark `awsConfig` as "may be null", and issue warnings at [L88](https://github.com/Brightspace/bmx/pull/454/files#diff-a6eaa1b3c5a510cd785801e0c37dfeb64b19bdb9263596f8ac5e9ed45bc9eba5R88) and below.

We could use the null-forgiving operator `!`, but it doesn't make our assumptions and intent clear:
* Why can we assume `awsConfig` isn't null in these places?
* When does this assumption hold?

By using `Debug.Assert`, we're effectively making clarifying statements like
* `foo` isn't null inside this if/else/for/while block
* `bar` isn't null once this above statement is executed successfully

We can also avoid sprinkle `!` in many different places.

---

The fact that this is also a runtime assertion in debug builds is more or less a side effect that I don't care that much. I don't think we should make it actually asserts at runtime when `BMX_DEBUG` is set either.

Reply:
We've used this technique to document our assumptions in code before, e.g.

* https://github.com/Brightspace/instance-metadata-store/blob/75ad7c513c12c6898d8e8397292ba67bc56cc9d0/api/src/D2L.InstanceMetadataStore.Api/CacheBuster.cs#L74

* https://github.com/Brightspace/instance-metadata-store/blob/75ad7c513c12c6898d8e8397292ba67bc56cc9d0/worker/src/D2L.InstanceMetadataStore.Worker/ConfigEtl/Extractor/ConfigExtractorProcessor.cs#L80

It's also used by the .NET runtime libraries, e.g.

* https://github.com/dotnet/runtime/blob/e1efa6b633d01c00ec019a8e222b2ebdfa593bcb/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/RedirectHandler.cs#L34

---

These are assumptions that should always hold true. If there's a real non-negligible chance that they could be false, then they should become real runtime assertions like
```csharp
if ( !assumption ) {
	throw new InvalidOperationException( "bad condition" );
}
```